pkg/proxy/proxy.go

```diff
@@ -185,32 +205,81 @@ func (p *Proxy) forwardRequest(host string, w http.ResponseWriter, req *http.Req
 		// If the client sent multiple XFF headers, flatten them.
 		proxyReq.Header.Set(xff, strings.Join(previous, ", "))
 	}
-	proxyReq.URL.Host = host
 	proxyReq.URL.Scheme = "https"
+	attempts := 0
 
-	log.Debug("Forwarding request to %s", proxyReq.URL.String())
-	client := http.Client{}
-	resp, err := client.Do(proxyReq)
-	if err != nil {
-		if urlErr, ok := err.(*url.Error); ok && urlErr.Timeout() {
-			writeJSONError(w, http.StatusGatewayTimeout, urlErr)
-		} else {
+	for {
+		proxyReq.URL.Host = acct.Host
+		log.Debug("Forwarding request to %s", proxyReq.URL.String())
+		client := http.Client{}
+		result, err := client.Do(proxyReq)
+
+		if err != nil {
+			if urlErr, ok := err.(*url.Error); ok && urlErr.Timeout() {
+				writeJSONError(w, http.StatusGatewayTimeout, urlErr)
+			} else {
+				writeJSONError(w, http.StatusBadGateway, err)
+			}
+			return
+		}
+
+		body := make([]byte, MaxResponseLength+1)
+		body, err = inet.ReadWithContext(ctx, result.Body, body)
+		result.Body.Close()
+
+		if err != nil {
 			writeJSONError(w, http.StatusBadGateway, err)
+			return
 		}
-		return
-	}
-	defer resp.Body.Close()
 
-	for _, hdr := range connectionHeaders {
-		resp.Header.Del(hdr)
-	}
-	outHeader := w.Header()
-	for name, value := range resp.Header {
-		outHeader[name] = value
-	}
+		if len(body) == MaxResponseLength+1 {
+			writeJSONError(w, http.StatusBadGateway, protocol.NewError("response exceeds maximum length", true, true))
+			return
+		}
+
+		if result.StatusCode == http.StatusOK {
+			for _, hdr := range connectionHeaders {
+				result.Header.Del(hdr)
+			}
+			outHeader := w.Header()
+			for name, value := range result.Header {
+				outHeader[name] = value
+			}
+
+			w.WriteHeader(result.StatusCode)
+			w.Write(body)
+			return
+		} else if result.StatusCode == http.StatusMisdirectedRequest && result.Header.Get("Alt-Svc") != "" {
+			altSvc := result.Header.Values("Alt-Svc")
+			idx := slices.IndexFunc(altSvc, func(str string) bool { return strings.HasPrefix(str, h2Prefix) })
+			if idx == -1 {
+				writeJSONError(w, result.StatusCode, err)
+				return
+			}
 
-	w.WriteHeader(resp.StatusCode)
-	io.Copy(w, resp.Body)
+			altHost := altSvc[idx][len(h2Prefix):]
+			log.Debug("Received HTTP Status 421. Updating server URL to %s", altHost)
+			acct.Host = altHost
+			p.updateDomainForSubject(acct.Subject, acct.Host)
+		} else {
+			writeJSONError(w, result.StatusCode, err)
+			return
+		}
+
+		attempts += 1
+		if attempts == MaxRetries {
```

Comment:
Nit: This would be more accurately named MaxAttempts. If MaxRetries = 1, then there aren't any retries, just the initial attempt.